pkg/rng/entropy.go

```
@@ -0,0 +1,168 @@
+	// in seconds
+	EntropyFeedTime time.Duration = 10
+	// RandomDevice is the linux random device
+	RandomDevice = "/dev/random"
+	// HwRandomDevice is the linux hw random device
+	HwRandomDevice = "/dev/hwrng"
+)
+
+// Can be extended but keep in mind to priorize
+// more secure random sources like hw random over
+// timer, jitter based mechanisms. Zero is the highest
+// priority.
+// <rng-name> : <priority>
+var trngList = map[string]int{
+	"tpm-rng":        0,
+	"intel-rng":      1,
+	"amd-rng":        1,
+	"timeriomem-rng": 2,
+}
+
+// Searches for available True Random Number Generator
+// inside the kernel api and sets the most secure on if
+// available which seeds /dev/hwrng
+func setAvailableTRNG() (bool, error) {
+	var (
+		currentRNG    string
+		availableRNGs []string
+		rngs          []string
+	)
+
+	currentFileData, err := ioutil.ReadFile(HwRandomCurrentFile)
+	if err != nil {
+		return false, err
+	}
+	currentRNG = string(currentFileData)
+
+	availableFileData, err := ioutil.ReadFile(HwRandomAvailableFile)
+	if err != nil {
+		return false, err
+	}
+	availableRNGs = strings.Split(string(availableFileData), " ")
+
+	for key, value := range trngList {
+		if key == currentRNG && value == 0 {
+			return true, nil
+		}
+	}
+
+	for _, rng := range availableRNGs {
+		for key := range trngList {
+			if rng == key {
+				rngs = append(rngs, key)
+			}
+		}
+	}
+
+	if len(rngs) <= 0 {
+		return false, nil
+	}
+
+	ioutil.WriteFile(HwRandomCurrentFile, []byte(rngs[0]), 0644)
+
+	return true, nil
+}
+
+// UpdateLinuxRandomness seeds random data from
+// /dev/hwrng into /dev/random based on a timer and
+// the entropy pool size
+// Usage:
+// go UpdateLinuxRandomness()
+func UpdateLinuxRandomness() {
+	good, err := setAvailableTRNG()
+	if !good {
+		log.Fatalf("Can't set available TRNG: %s", err.Error())
+		return
+	}
+
+	randomPoolSizeData, err := ioutil.ReadFile(RandomPoolSizeFile)
+	if err != nil {
+		log.Fatalf("Can't random poolsize data: %s", err.Error())
```

you can use `%v` and just `err` (i.e. `log.Printf("error: %v", err)`).

Additionally, `log.Fatal` is dangerous outside of a `main` because it calls `exit` straight away, and `defer` statements are not honoured. You can either return an error, or use `log.Panic` if you really want to break the program. I'd rather return an error because it lets the caller decide how to recover from this condition

---

Done